python/ray/serve/_private/autoscaling_state.py

```diff
@@ -0,0 +1,355 @@
+import logging
+import time
+from dataclasses import dataclass
+from typing import Dict, List, Optional, Set
+
+from ray.serve._private.autoscaling_policy import AutoscalingPolicyManager
+from ray.serve._private.common import (
+    DeploymentHandleSource,
+    DeploymentID,
+    ReplicaID,
+    TargetCapacityDirection,
+)
+from ray.serve._private.constants import (
+    RAY_SERVE_COLLECT_AUTOSCALING_METRICS_ON_HANDLE,
+    RAY_SERVE_MIN_HANDLE_METRICS_TIMEOUT_S,
+    SERVE_LOGGER_NAME,
+)
+from ray.serve._private.deployment_info import DeploymentInfo
+
+logger = logging.getLogger(SERVE_LOGGER_NAME)
+
+
+@dataclass
+class HandleRequestMetric:
+    actor_id: str
+    handle_source: DeploymentHandleSource
+    queued_requests: float
+    running_requests: Dict[ReplicaID, float]
+    timestamp: float
+
+    @property
+    def total_requests(self) -> float:
+        return self.queued_requests + sum(self.running_requests.values())
+
+    @property
+    def is_serve_component_source(self) -> bool:
```

Comment:
Could you add a comment explaining when this evaluates to `False`?

It looks like [this is used to track](https://github.com/ray-project/ray/pull/44638/files#diff-47e03417ac344cec3fbcd930945c85aef89a773e22b59e314b6523c0f133a5dfR170-R175) whether the handle is on a dead replica/proxy. Why do we need this in addition to checking only the `actor_id`? Is it possible to pull the full check into this function and use a more descriptive name (e.g. `is_on_living_actor`)?

Reply:
This is necessary because we remove handles that are on dead Serve actors, by checking if the actor id is in the set of alive Serve actors. We only track the set of Serve replica and proxy actors though, so we can only do this for handles that came from Serve replicas/proxies (if it's on any other worker, we'd always be dropping the metric since those workers aren't tracked in the controller). I've added a comment, hopefully that clears it up.